### ISSUE TITLE
[BUG] Partitioned activation checkpointing does not reduce memory usage

### ISSUE THREAD

**Issue by andrasiani** [microsoft/DeepSpeed#3779]
Hi, I have an issue with partitioned activation checkpointing.
When I run activation_checkpointing withoout partitioning I get the same memory usage as in the case of partitioning enabled.
deepspeed/runtime/activation_checkpoing/checkpointing.py 

I checked and partition_activations() method shards the tensor correcly,
the backward all_gather method also seems to work correctly, 
the get_partitioned_activations_for_backward seems to only save partitions.

Somewhere I suspect that the whole checkpointed layer is stored in memory after forward until backward pass, instead of just a partition of it.

This is the checkpointing script I use:
[checkpointing_ custom.txt](https://github.com/microsoft/DeepSpeed/files/11815533/checkpointing_.custom.txt)

and I call it like this:
# mpu object is set to None in pytorch lightning - this is a pytorch lightning bug I think - inside my custom script I initialize an mpu object:
            deepspeed.checkpointing.configure(
                mpu_=None,  
                partition_activations=checkpoint_config.get("partition_activations"),
                contiguous_checkpointing=checkpoint_config.get("contiguous_memory_optimization"),
                checkpoint_in_cpu=checkpoint_config.get("cpu_checkpointing"),
                profile=checkpoint_config.get("profile"),
            )

**Comment by RoyJames** [comment]
Same issue. Also tried the solution by @andrasiani which works. Let's hope either DeepSpeed or Pytorch lightning has better support for this, otherwise partitioned checkpointing is totally unusable.

**Comment by RoyJames** [comment]
> @RoyJames thanks for trying that out. Do you see any benefit after the "fix" above, any memory reduction before and after enabling activation partitioning?

I stepped into relevant partitioning functions and can confirm that your fix successfully addressed the MPU issue. But there's still one more bug that prevents me from getting proper memory saving as described in https://github.com/microsoft/DeepSpeed/issues/3857

**Comment by suuperstone** [comment]
Deepspeed only provides  Pipeline parallelism (PP), and using Deepspeed PP is incompatible with zero-2 and zero-3.
- ref : https://deepspeed.readthedocs.io/en/latest/pipeline.html

While the official documentation suggests using activation partitioning with zero-3 + MP, in reality, you cannot use activation partitioning alongside zero-3.

Furthermore, zero-3 calculates activations independently for each GPU in a sub-batch, making activation partitioning itself meaningless.

In conclusion, if you're using the DeepspeedStrategy in PyTorch Lightning, 
- applying activation partitioning doesn't offer much significance.
- applying activation partitioning with zero is impossible.

I'm also studying this issue extensively, and I would like to discuss it further if possible

**Comment by andrasiani** [comment]
Hi @LogicBaron, I wonder why they have the partitioned activations flag at all then?
Do you know any other framework that makes activation partitioning possible? FSDP and Deepspeed are not making this possible at the moment.
Correct me if I'm wrong.
Also, would be nice to have some feedback from deepspeed developers, they are very silent on this topic.

**Comment by LogicBaron** [comment]
@andrasiani 

I also failed to use Partitioned Activation Checkpointing with deepspeed, and it was difficult when I tried to implement it.

However, I found that just activation checkpointing was effective enough, but is it difficult to train your model with only activation checkpointing?

**Comment by tjruwase** [comment]
@andrasiani, @LogicBaron, your observations regarding partitioned activation checkpointing are correct. Partitioned activation checkpointing is only beneficial for tensor-parallelism (TP) to avoid the checkpointing duplication on the TP ranks. Please see 4.2 of [zero paper](https://arxiv.org/pdf/1910.02054.pdf) for more details 
![image](https://github.com/microsoft/DeepSpeed/assets/4271600/d4874e85-a08f-4dd4-a788-b7fb486f2ac7)

Apologies for the delayed response, but I believe this issue can be closed with this clarification.

**Comment by andrasiani** [comment]
Thank you, @tjruwase, however this still does not explain why is there a partition activations flag in lightning if it cannot be used.
Could you provide a scenario/config where that flag works and the activations are partitioned?

**Comment by tjruwase** [comment]
@andrasiani, unfortunately I don't know why the flag is in lightning code, since they wrote the integration code. 

Megatron-LM's tensor parallelism is a scenario where this flag helps. Unfortunately, our [codebase](https://github.com/microsoft/DeepSpeedExamples/blob/staging-zero3-release/Megatron-LM-v1.1.5-ZeRO3/examples/ds_pretrain_gpt2-zero3.sh) that demonstrated this feature is now deprecated. You can see the flag activated [here](https://github.com/microsoft/DeepSpeedExamples/blob/staging-zero3-release/Megatron-LM-v1.1.5-ZeRO3/examples/ds_pretrain_gpt2-zero3.sh#L116-L118). There is no guarantee that this deprecated codebase will work with latest deepspeed, instead we recommend using the replacement [Megatron-DeepSpeed](https://github.com/microsoft/Megatron-DeepSpeed). Although, the flag is not available in the new codebase, this a decent [script](https://github.com/microsoft/Megatron-DeepSpeed/blob/main/examples_deepspeed/azure/run-175b.sh) that you could add it to for your exploration. Remember to [enable tensor parallelism](https://github.com/microsoft/Megatron-DeepSpeed/blob/a4f807982d7e3552d2d7f1ff2f37a89efd529759/examples_deepspeed/azure/run-175b.sh#L18) by setting TP > 1.

**Comment by andrasiani** [comment]
@tjruwase is it possible to apply zero-r or any kind of activation partitioning for a gpt2 model using pytorch lightning? That`s all I am interested in.
